src/main/java/pink/zak/minestom/towerdefence/queue/MobQueue.java

```diff
@@ -0,0 +1,129 @@
+package pink.zak.minestom.towerdefence.queue;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.stream.Collectors;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextDecoration;
+import net.minestom.server.MinecraftServer;
+import net.minestom.server.item.ItemStack;
+import net.minestom.server.item.Material;
+import net.minestom.server.item.metadata.BundleMeta;
+import net.minestom.server.timer.Task;
+import net.minestom.server.timer.TaskSchedule;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import pink.zak.minestom.towerdefence.api.event.player.PlayerQueueUpdateEvent;
+import pink.zak.minestom.towerdefence.game.MobHandler;
+import pink.zak.minestom.towerdefence.model.mob.config.EnemyMob;
+import pink.zak.minestom.towerdefence.model.user.GameUser;
+import pink.zak.minestom.towerdefence.utils.Result;
+
+public final class MobQueue {
+
+    private static final int MAX_QUEUE_TIME = 45_000; // 45 seconds
+    private static final @NotNull ItemStack BASE_QUEUE_ITEM = ItemStack.builder(Material.BUNDLE)
+            .displayName(Component.text("Current Queue", NamedTextColor.YELLOW).decoration(TextDecoration.ITALIC, false))
+            .build();
+
+    private final @NotNull Queue<QueuedEnemyMob> queue = new ConcurrentLinkedQueue<>();
+
+    private final @NotNull MobHandler mobHandler;
+    private final @NotNull GameUser user;
+
+    private final @NotNull Task task;
+    private long nextSpawnTime = 0;
+    private @Nullable QueuedEnemyMob currentlySpawning;
+
+    public MobQueue(@NotNull MobHandler mobHandler, @NotNull GameUser user) {
+        this.mobHandler = mobHandler;
+        this.user = user;
+
+        this.task = MinecraftServer.getSchedulerManager().buildTask(this::tick)
+                .repeat(TaskSchedule.nextTick())
+                .schedule();
+    }
+
+    public void unregister() {
+        this.task.cancel();
+    }
+
+    private void tick() {
+        // check if we have a mob to send
+        if (this.currentlySpawning == null) {
+            this.currentlySpawning = this.queue.poll();
+            if (this.currentlySpawning == null) return;
+            this.nextSpawnTime = System.currentTimeMillis() + this.currentlySpawning.mob().getSendTime();
+        }
+
+        // check if we can send the mob
+        if (System.currentTimeMillis() < this.nextSpawnTime) return;
+
+        // send the mob
+        this.mobHandler.spawnMob(this.currentlySpawning, this.user);
+        this.currentlySpawning = null;
+
+        // broadcast the queue event
+        this.callUpdateEvent();
+    }
```

Comment:
The mob is removed from the queue at the start of their spawn timer. This means the mob is technically no longer in the queue way before it even spawns. It should probably be removed from the queue after it has spawned.

Reply:
This does not seem to be the case when testing.. must be too late in the evening.